### ISSUE TITLE
`suggest_memory_format` has ambiguity & cannot represent intended layout format for corner cases

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#24090]
## 🐛 Bug

`suggest_memory_format` has ambiguous meaning for two cases:
1. tensor with NCHW where C = 1.
   we could use stride of C as a hint to tell the intended memory format.
2. tensor with NCHW where H == W == 1.
   there's no way to identify the intended memory format from strides.

The problem with this is: There's no proper way to correctly represent desired memory format for tensor with shape NC11. (yes, we can always hack it by utilizing/special casing stride for size 1 dimension. hence the emphasis on *proper*)

This would impact our cudnn setup. For 1x1 convolution, we have filter size with KC11. We would want to specify memory format properly for this case as it would impact the algo we get.

**Comment by jjsjann123** [comment]
Keeping track of this by cc'ing #23403 #19092

**Comment by apaszke** [comment]
Well in those cases both layouts are actually equivalent, so the potential ambiguity should be handled entirely by the backends.

**Comment by jjsjann123** [comment]
Realized that the phrasing & the example was done poorly that kinda distracted the issue I was complaining.

The problem here is not about a specific layer malfunctioning. But the capability of our NHWC proposal to correctly represent the memory format intended by the user.

And I realize that I made a mistake earlier. Case 1 (N1HW) and case 2 (NC11) could be worked out using our existing strategy separately.
What needed to be done here is, we shouldn't ignore strides for size 1 dimensions inside our `compute_strides_like_channels_last()`: https://github.com/pytorch/pytorch/blob/6e4a83ab577887fc3c692e106433a121c3165130/c10/core/TensorImpl.cpp#L115-L130
We can use those strides to hint the intended memory layout.

Then for cases with (N111) input tensor, there's nothing we can do, but that probably also doesn't really matter in real life.

I'll issue a PR to fix this.

**Comment by apaszke** [comment]
I don't think it's a mistake of our system, again --- those layouts are equivalent in those cases. It's a matter of cuDNN not being smart enough to know that 1x1 NCHW filters actually work fine as 1x1 NHWC filters.

**Comment by jjsjann123** [comment]
I agree that backends should handle strides properly and be able to . CUDNN check actually checks the strides other than the descriptor so hopefully it will be fine.

But the issue is also relevant in our front end,
how do user script intended behavior and how does PyTorch handle/deduce the right thing to do.

Taking that conv example again here.
```
import torch


x = torch.randn(4, 1, 20, 20, device="cuda")
x = x.contiguous(memory_format=torch.channels_last)

x_ref = torch.randn(4, 4, 20, 20, device="cuda")
x_ref = x_ref.contiguous(memory_format=torch.channels_last)

conv = torch.nn.Conv2d(1, 20, 1).cuda()
conv_ref = torch.nn.Conv2d(4, 20, 1).cuda()

out = conv(x)
out_ref = conv_ref(x_ref)

print(out.stride())
print(out_ref.stride())
```

`out` will be in NHWC layout while `out_ref` will be in NCHW.
It has nothing to do with the backend.  ATen allocates output tensor by inspecting `suggest_memory_format()` on the input `x` (arguably we can also check the weight memory layout, but it doesn't help the case here since `conv.weight.size() == (20,1,1,1)`).

This could be easily fixed with the hinting stride PR that I was proposing.
